win32ss/gdi/eng/mouse.c

```diff
@@ -787,11 +792,22 @@ GreMovePointer(
     /* Check if we need to move it */
     if(pdc->ppdev->SafetyRemoveLevel == 0)
     {
+        SURFOBJ* pso = &pdc->ppdev->pSurface->SurfObj;
+        PFN_DrvMovePointer pfnMovePointer;
+
         /* Store the cursor exclude position in the PDEV */
         prcl = &pdc->ppdev->Pointer.Exclude;
 
         /* Call Eng/Drv function */
-        pdc->ppdev->pfnMovePointer(&pdc->ppdev->pSurface->SurfObj, x, y, prcl);
+        if (pdc->ppdev->flFlags & PDEV_HARDWARE_POINTER)
+            pfnMovePointer = pdc->ppdev->pfnMovePointer;
+        else
+            pfnMovePointer = EngMovePointer;
+        pfnMovePointer(pso, x, y, prcl);
+
+        /* If panning device, and we're not hiding the cursor, call panning driver */
+        if (pdc->ppdev->devinfo.flGraphicsCaps & GCAPS_PANNING && y != -1)
+            pdc->ppdev->pfnMovePointer(pso, x, y - pso->sizlBitmap.cy, NULL);
```

Comment:
So here it's really a direct call of `pfnMovePointer`, instead of the selection as done above?
I suppose then that it's not possible that both the following conditions:
`pdc->ppdev->flFlags & PDEV_HARDWARE_POINTER`
and
`pdc->ppdev->devinfo.flGraphicsCaps & GCAPS_PANNING && y != -1`
can be TRUE at the same time? (otherwise it would look like the `pdc->ppdev->pfnMovePointer` function would be called twice)

Reply:
I'll be happy to drop changes in mouse.c, if you find when/how pfnMovePointer is called.
Note that MouseSafetyOnDrawStart/MouseSafetyOnDrawEnd also hide/show the cursor to probably hide another problem.

Documentation doesn't say that both calls are exclusive, but I'm not sure. PDEV_HARDWARE_POINTER can only be set if both DrvMovePointer and DrvSetPointerShape are available.
The first call (y >= 0) will be to position the hot spot of the hardware pointer, while the second one (y < 0) is to notify cursor current position for panning driver (GCAPS_PANNING flag).

Reply:
I would then advice to add a bit more comments in the code that explain that (so that we don't forget about it later on)

Reply:
Done